### DISCUSSION TITLE
Access control over query according to filter

### DISCUSSION THREAD

**Discussion by ConductedClever** [keystonejs/keystone#7592] [Questions] [answered]
Hi,

According to [this documentation](https://keystonejs.com/docs/apis/access-control) on keystonejs access control, there are three levels to check user access, operation, filter and item. 

There are user submissions on a model named ResearchRegistration. Each submission is unique on the triple of nationalCode, email and phoneNumber. 

Now I want to limit the clients to read the records only if they can provide the correct triple. So if the client provide wrong filter no records will be returned but if the client sends a query without any filter I can't reject the request as I don't know the user sent filter on access step. I want to allow the users to query the list but enforce them to provide these three filters (nationalCode, email and phoneNumber) and reject the other requests that excludes either of the filters mentioned.

```typescript
...
export default list({
...
  access: {
    operation: {
      create: ({ session, context, listKey, operation }) => {
        return isAdminOrContentManager(session) || checkIfCaptchaIsPassed((<any>context).req);
      },
      query: ({ session, context, listKey, operation }) => {
        return isAdminOrContentManager(session); // HERE
      },
      update: ({ session, context, listKey, operation }) => {
        return isAdminOrContentManager(session) || checkIfCaptchaIsPassed((<any>context).req); //&& checkIfRequestSchemaIsValid((<any>context).req));
      },
      delete: ({ session, context, listKey, operation }) => {
        return isAdminOrContentManager(session);
      },
    },
    filter: {
      query: ({ session, context, listKey, operation }) => {
        return true; // or HERE
      },
    },
  },
  fields: {
...
    email: text({ validation: { isRequired: true } }),
    nationalCode: text({ validation: { isRequired: true } }),
    phoneNumber: text({ validation: { isRequired: true } }),
...
  },
});
```
In the commented lines (with HERE) I need to know whether the filter has been sent by user so that I can make a decision about allowing or denying their request.

**Comment by emmatown** [comment, marked as answer]
The functions in `.access.operation.*`  and `.access.filter.*` are not provided the parsed GraphQL query or parameters as they can be too dynamic and difficult to understand.

For use cases like this, you could limit access control to whoever who should be able to fetch the items, and then use `extendGraphqlSchema` to write a custom query that meets your exact expectations.

The custom query could then accept the exact parameters you need, and it could adhere to the access control you have on the `context` provided to the custom query.

**Reply by ConductedClever** [reply]
Another scenario where I need to know the parsed GraphQL query in, is one that I call it mark as read. I have made a model named `JobApplication` and I want to let the admin know if he/she had opened the job application before or not. So when they open the list of job applications (query all) in the keystone admin area I want to do nothing and then if one is opened (query on by ID) mark it as read (change the value of flag `isRead` to true). So in the `access.filter` I need to know the parsed GraphQL query that the user has been sent to infer that the query is a list query or item query and also have the ID to patch the read flag on it.  
So providing the parsed GraphQL query in `.access.operation.*` and `.access.filter.*` will help me so much and implementing by using `extendGraphqlSchema` will be more complicated and less maintainable for me.

**Reply by dcousens** [reply]
Thanks for the extended explanation @ConductedClever - it helps us understand the problems users face and whether or not the API's are suitable or not
